### PR TITLE
Fix section reuse

### PR DESCRIPTION
**Pull request by kd-11**
Probable fix for https://github.com/RPCS3/rpcs3/issues/4298
- Returns things to the old way of handling false positives in the texture cache.

Thanks to @ruipin for debugging the issue. Took us a whole week to find my silly mistakes 😝. Hopefully this fixes flickering in other games like https://github.com/RPCS3/rpcs3/issues/4857

### CODE REVIEW

**Comment by ruipin** [comment]
Mistakes happen ;)
I'm curious if this actually fixes more games than P5!

Once this gets merged I'll see about push some of the sanity checks/assertions I wrote to debug this, so we can avoid more texture cache issues in the future.

**Comment by Dani1977** [comment]
I try this PR, white flickering are gone , but black flickering (black screen) still present this is much shorter and more rare than before .

I try Nexus and Boletaria 1-1, ask me if you need more information and tests .

Regards 

[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2302749/RPCS3.log.gz)

**Comment by kd-11** [comment]
You can try with async disabled just in case there are issues resulting from that. Its unlikely though, likely some other unrelated bug that will need separate investigation.

**Comment by Dani1977** [comment]
immagine

immagine

[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2303230/RPCS3.log.gz)

with disable  async shaders (on already visited level so no caching shaders) looks like the same with enable; but I notice this thing (enable/disable async is the same): 
when you dodge on a barrel or broke some environments, the screen flick (Boletaria). 

Hope that can help you 
Regards

**Comment by fifazalata** [comment]
Could this be related to WCB?
Since the flickering was happening with NieR too and it needs WCB.

**Comment by kd-11** [comment]
Nier bug predates the memory mirror patches. Likely not related.

**Comment by Parthorisian** [comment]
Non TSX CPU. Shaders cache has been cleared. No any flickering noticed+large performance boost(ASMJIT Recompiler) similar to pr4935. 
Settings: standard except WCB+Use GPUTexture scaling - are on. Played around 1-1,5 hours (almost 600 pipeline objects compiled)
P.S. a)May be something like that sometimes for Nexus+World4-2, but hard to notice and i think it was just popups of uncompiled shaders. Could'nt notice anything on 2nd boot
      b)I hope infinite loading screen (for ASMJIT Recompiler) which finished my tests, unrelated to this PR
[DemonsSouls_RPCS3.zip](https://github.com/RPCS3/rpcs3/files/2307003/DemonsSouls_RPCS3.zip)

**Comment by kamer1337** [comment]
Tony Hawk's Project 8

> F {PPU[0x1000000] Thread (main_thread) [0x009f1edc]} class std::runtime_error thrown: Unreachable
> (in file c:\projects\rpcs3\rpcs3\emu\rsx\vk\../Common/texture_cache.h:614)

[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2307240/RPCS3.log.gz)

**Comment by kd-11** [comment]
Ignore the assertions for now, I'll update when its ready for testing again. Currently restructuring some internals so things are still buggy.

**Comment by kd-11** [comment]
Should be safe to test again.

**Comment by dukenukemx** [comment]
Been using #5013 and it still freezes but it will unfreeze if I give it some time.  It has done this twice.  I still get the flicker once in a while too.  In both cases the RPCS3 main window is not accessible.  It isn't usable cause once this happens it freezes again in seconds.

File: rpcs3/Emu/RSX/Common/texture_cache.h [hunk 1]
```diff
@@ -564,7 +565,131 @@ namespace rsx
...
+			result.reserve(marked_sections.size());
+
+			for (const auto &section : marked_sections)
+			{
+				result.push_back(section->get_protected_range());
+			}
+
+			for (const auto &dirty : sections_to_exclude)
+			{
+				const auto intersect_test = dirty->get_protected_range();
```
Comment (ruipin) [inline rpcs3/Emu/RSX/Common/texture_cache.h]: I suggest changing this to `exclude_range`, `exclude_start` and `exclude_end` for readability.

File: rpcs3/Emu/RSX/Common/texture_cache.h [hunk 2]
```diff
@@ -564,7 +565,131 @@ namespace rsx
 			return true;
 		}
 
-		//Get intersecting set - Returns all objects intersecting a given range and their owning blocks
+		std::vector<std::pair<u32, u32>> subtractive_intersect(std::vector<section_storage_type*> marked_sections, std::vector<section_storage_type*> sections_to_exclude)
+		{
+			std::vector<std::pair<u32, u32>> result;
+			bool intersection_exists = false;
```
Comment (ruipin) [inline rpcs3/Emu/RSX/Common/texture_cache.h]: Unused variable

**Comment by ruipin** [comment]
I have tested this commit (or to be exact https://github.com/ruipin/rpcs3/commit/4e44fbd18d117b816d14022df35d2f41646247d0 which is a version of this commit with the review fixes applied and a ton of debug checks and assertions) on both P5 and DeS for about 1 hour each, and could not reproduce any issues. Just make sure to fix the bug I reported in one of the review comments.

File: rpcs3/Emu/RSX/Common/texture_cache.h [hunk 3]
```diff
@@ -564,7 +565,131 @@ namespace rsx
...
+					// At least one intersection exists
+					intersection_exists = true;
+
+					// Split the range if possible
+					if (range_start >= memory_base && range_end <= memory_end)
+					{
+						// Cannot be salvaged
+						this_range = { 0, 0 };
+					}
+					else if (memory_end > range_start && memory_end < range_end)
```
[the remark this answers is not in the document]
Comment (kd-11) [reply rpcs3/Emu/RSX/Common/texture_cache.h]: its inverse of your test, its checking if `memory_end` overlaps `{range_start, range_end}`. Will refactor for readability.

File: rpcs3/Emu/RSX/Common/texture_cache.h [hunk 4]
```diff
@@ -325,8 +325,9 @@ namespace rsx
 		struct thrashed_set
 		{
 			bool violation_handled = false;
-			std::vector<section_storage_type*> sections_to_flush; //Sections to be flushed
-			std::vector<section_storage_type*> sections_to_unprotect; //These sections are to be unpotected and discarded by caller
+			std::vector<section_storage_type*> sections_to_flush; // Sections to be flushed
+			std::vector<section_storage_type*> sections_to_unprotect; // These sections are to be unpotected and discarded by caller
+			std::vector<section_storage_type*> sections_to_exclude; // These sections are to be repotected and discarded by caller (false positives)
```
Comment (Margen67) [inline rpcs3/Emu/RSX/Common/texture_cache.h]: *protected

**Pull request merged**